modemController.go

```diff
@@ -0,0 +1,149 @@
+/*
+modemd - Communicates with USB modems
+Copyright (C) 2019, The Cacophony Project
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"os/exec"
+	"time"
+
+	"periph.io/x/periph/conn/gpio"
+	"periph.io/x/periph/conn/gpio/gpioreg"
+)
+
+type ModemController struct {
+	StartTime         time.Time
+	Modem             *Modem
+	ModemsConfig      []ModemConfig
+	TestHosts         []string
+	TestInterval      int
+	PowerPin          string
+	InitialOnTime     int
+	FindModemTime     int // Time in seconds after USB powered on for the modem to be found
+	ConnectionTimeout int // Time in seconds for modem to make a connection to the network
+	PingWaitTime      int
+	PingRetries       int
+	RequestOnTime     int // Time the modem will stay on in seconds after a request was made
+
+	lastOnRequestTime time.Time
+}
+
+func (mc *ModemController) NewOnRequest() {
+	log.Println("stay on request")
+	mc.lastOnRequestTime = time.Now()
+}
+
+func (mc *ModemController) FindModem() bool {
+	startTime := time.Now()
+	for {
+		for _, modemConfig := range mc.ModemsConfig {
+			cmd := exec.Command("lsusb", "-d", modemConfig.VendorProduct)
+			if err := cmd.Run(); err == nil {
+				mc.Modem = NewModem(modemConfig)
+				return true
+			}
+		}
+		if timeoutCheck(startTime, mc.FindModemTime) {
+			return false
+		}
+		time.Sleep(time.Second)
+	}
+}
+
+func (mc *ModemController) SetModemPower(on bool) error {
+	pin := gpioreg.ByName(mc.PowerPin)
+	if on {
+		if err := pin.Out(gpio.High); err != nil {
+			return fmt.Errorf("failed to set modem power pin high: %v", err)
+		}
+	} else {
+		if err := pin.Out(gpio.Low); err != nil {
+			return fmt.Errorf("failed to set modem power pin low: %v", err)
+		}
+	}
+	return nil
+}
+
+func (mc *ModemController) CycleModemPower() error {
+	if err := mc.SetModemPower(false); err != nil {
+		return err
+	}
+	time.Sleep(time.Second * 5)
```

Comment:
every call to `SetModemPower(false)` is followed by a 5s sleep so perhaps the sleep should be part of `SetModemPower`